src/cockpit/router.py

```diff
@@ -26,12 +27,54 @@
 logger = logging.getLogger(__name__)
 
 
+class ExecutionQueue(collections.deque):
+    """Temporarily delay calls to a given set of class methods.
+
+    Functions by replacing the named function at the instance __dict__
+    level, effectively providing an override for exactly one instance
+    of `method`'s object.
+    Queues the invocations.  Run them later with .run(). Reverse this call
+    by deleting the named methods from the instance.
+    """
+    def __init__(self, methods):
+        super().__init__()
+        self.methods = methods
+
+        for method in self.methods:
+            self._wrap(method)
+
+    def _wrap(self, method):
+        # NB: this function is stored in the instance dict and therefore
+        # doesn't function as a descriptor, isn't a method, doesn't get bound,
+        # and therefore doesn't receive a self parameter
+        setattr(method.__self__, method.__func__.__name__, lambda *args: self.append((method, args)))
+
+    def run(self):
+        for method, args in self:
+            print(method, args)
+            method(*args)
+
+        for method in self.methods:
+            delattr(method.__self__, method.__func__.__name__)
+
+
 class Endpoint:
     router: Router
+    queue: Optional[ExecutionQueue] = None
 
     def __init__(self, router: Router):
         self.router = router
 
+    def freeze_endpoint(self):
```

Comment:
I tried the opposite here, `assert self.queue is None`, for preventing double freezing. That seems fine.

However, I wonder if `queue` should really be a class variable instead of an instance one? Doesn't that mean that all channels (which are Endpoints, right?) share a single queue?

Reply:
This is the only outstanding thread here. Adding an assertion here would be nice, but also not super-important. I'm happy with this PR once it gets green. Thanks!